numpy/core/src/multiarray/item_selection.c

```diff
@@ -787,255 +787,303 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
-    }
-
-    size = it->size;
-    N = PyArray_DIMS(op)[axis];
-    elsize = PyArray_DESCR(op)->elsize;
-    astride = PyArray_STRIDES(op)[axis];
 
-    needcopy = !(PyArray_FLAGS(op) & NPY_ARRAY_ALIGNED) ||
-                (astride != (npy_intp) elsize) || swap;
     if (needcopy) {
-        char *buffer = PyDataMem_NEW(N*elsize);
+        buffer = PyDataMem_NEW(N * elsize);
         if (buffer == NULL) {
             goto fail;
         }
+    }
 
-        while (size--) {
-            _unaligned_strided_byte_copy(buffer, (npy_intp) elsize, it->dataptr,
-                                         astride, N, elsize);
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
-            }
-            if (part == NULL) {
-                if (sort(buffer, N, op) < 0) {
-                    PyDataMem_FREE(buffer);
-                    goto fail;
+    while (size--) {
+        char *bufptr = it->dataptr;
+
+        if (needcopy) {
+            if (hasrefs) {
+                /*
+                 * For dtype's with objects, copyswapn Py_XINCREF's src
+                 * and Py_XDECREF's dst. This would crash if called on
+                 * an unitialized buffer, or leak a reference to each
+                 * object if initialized.
+                 *
+                 * So, first do the copy with no refcounting...
+                 */
+                _unaligned_strided_byte_copy(buffer, elsize,
+                                             it->dataptr, astride, N, elsize);
+                /* ...then swap in-place if needed */
+                if (swap) {
+                    copyswapn(buffer, elsize, NULL, 0, N, swap, op);
                 }
             }
             else {
-                npy_intp pivots[NPY_MAX_PIVOT_STACK];
-                npy_intp npiv = 0;
-                npy_intp i;
-                for (i = 0; i < nkth; i++) {
-                    if (part(buffer, N, kth[i], pivots, &npiv, op) < 0) {
-                        PyDataMem_FREE(buffer);
-                        goto fail;
-                    }
-                }
+                copyswapn(buffer, elsize, it->dataptr, astride, N, swap, op);
             }
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
+            bufptr = buffer;
+        }
+        /*
+         * TODO: If the input array is byte-swapped but contiguous and
+         * aligned, it could be swapped (and later unswapped) in-place
+         * rather than after copying to the buffer. Care would have to
+         * be taken to ensure that, if there is an error in the call to
+         * sort or part, the unswapping is still done before returning.
+         */
+
+        if (part == NULL) {
+            ret = sort(bufptr, N, op);
+            #if defined(NPY_PY3K)
+            /* Object comparisons may raise an exception in Python 3 */
+            if (hasrefs && PyErr_Occurred()) {
```

Comment:
NPY_PY3K could be put into the if ()
also preprocessor code is usually not indented